# Berlin Food & Weekly Markets - Data Transformation

**Author:** Michael Wetzel  
**Date:** November 2025  
**Purpose:** Transform and combine multiple data sources for Berlin food markets database

This notebook performs:
1. Loading and combining multiple data sources
2. Data cleaning and standardization
3. Filtering (flea markets, past Christmas markets)
4. Spatial join to add neighborhood_id and district_id
5. Address completion using Nominatim
6. Export to CSV with NO missing values

## Setup and Imports

In [47]:
import pandas as pd
import geopandas as gpd # Install and import geopandas for spatial operations
from shapely.geometry import Point
from geopy.geocoders import Nominatim
from geopy.exc import GeocoderTimedOut, GeocoderServiceError
import time
import json
import re
from pathlib import Path
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')

## Configuration

In [48]:
# File paths
SOURCES_DIR = Path('food_markets/sources')
OUTPUT_FILE = 'food_markets/berlin_food_markets_clean.csv'

# District ID mapping (official 8-digit Berlin codes for foreign key constraint)
DISTRICT_MAPPING = {
    'Mitte': '11001001',
    'Friedrichshain-Kreuzberg': '11002002',
    'Pankow': '11003003',
    'Charlottenburg-Wilmersdorf': '11004004',
    'Spandau': '11005005',
    'Steglitz-Zehlendorf': '11006006',
    'Tempelhof-Schöneberg': '11007007',
    'Neukölln': '11008008',
    'Treptow-Köpenick': '11009009',
    'Marzahn-Hellersdorf': '11010010',
    'Lichtenberg': '11011011',
    'Reinickendorf': '11012012'
}

# Final schema including primary key and foreign keys
FINAL_COLUMNS = [
    'market_id',  # Primary key
    'market_name', 'market_type', 'address', 'postal_code', 'district',
    'neighborhood_id', 'district_id', 'opening_days', 'opening_hours',
    'operator', 'contact_email', 'website', 'accessibility',
    'latitude', 'longitude', 'data_source', 'notes'
]

print("Configuration loaded ✓")

Configuration loaded ✓


## Step 1: Load Data Sources

In [49]:
print("\n📥 Loading Data Sources")
print("-" * 70)

# Official GeoJSON sources (source of truth)
with open(SOURCES_DIR / 'weihnachtsmaerkte.geojson', 'r', encoding='utf-8') as f:
    christmas_data = json.load(f)
print(f"✓ Christmas markets: {len(christmas_data['features'])}")

with open(SOURCES_DIR / 'wochen-troedelmaerkte.geojson', 'r', encoding='utf-8') as f:
    weekly_data = json.load(f)
print(f"✓ Weekly/Flea markets: {len(weekly_data['features'])}")

# OSM data for gap filling
with open(SOURCES_DIR / 'OSM-berlin_markets.json', 'r', encoding='utf-8') as f:
    osm_data = json.load(f)
print(f"✓ OSM markets: {len(osm_data['elements'])}")

# Additional sources
df_wochenmarkt_de = pd.read_csv(SOURCES_DIR / 'wochenmarkt_deutschland.csv')
print(f"✓ Wochenmarkt Deutschland: {len(df_wochenmarkt_de)}")

with open(SOURCES_DIR / 'visitberlin_markets.txt', 'r', encoding='utf-8') as f:
    visitberlin_names = [line.strip() for line in f if line.strip()]
print(f"✓ VisitBerlin markets: {len(visitberlin_names)}")


📥 Loading Data Sources
----------------------------------------------------------------------
✓ Christmas markets: 47
✓ Weekly/Flea markets: 41
✓ OSM markets: 147
✓ Wochenmarkt Deutschland: 93
✓ VisitBerlin markets: 19


## Step 2: Helper Functions

In [50]:
def parse_german_date(date_str):
    """Parse German date format DD.MM.YYYY to datetime"""
    if not date_str or pd.isna(date_str):
        return None
    date_str = str(date_str).strip().split('\n')[0].strip()
    try:
        return datetime.strptime(date_str, '%d.%m.%Y')
    except:
        return None

def is_current_christmas_market(data):
    """Check if Christmas market is current (not from past years)"""
    bis_date = parse_german_date(data.get('bis', ''))
    if bis_date:
        # Market is current if end date is in the future or this season
        return bis_date >= datetime.now()
    # If no end date, assume it's current
    return True

def extract_geojson_to_df(geojson_data, market_type):
    """Extract GeoJSON properties.data to DataFrame"""
    records = []
    for feature in geojson_data['features']:
        data = feature['properties']['data'].copy()
        
        # Filter out past Christmas markets
        if market_type == 'christmas_market':
            if not is_current_christmas_market(data):
                continue  # Skip this market
        
        data['market_type'] = market_type
        data['data_source'] = 'berlin.de_official'
        records.append(data)
    return pd.DataFrame(records)

def standardize_columns(df):
    """Standardize column names"""
    mapping = {
        'name': 'market_name', 'bezeichnung': 'market_name',
        'strasse': 'address', 'plz_ort': 'postal_code', 'plz': 'postal_code',
        'bezirk': 'district', 'tage': 'opening_days', 'zeiten': 'opening_hours',
        'oeffnungszeiten': 'opening_hours', 'veranstalter': 'operator',
        'betreiber': 'operator', 'email': 'contact_email', 'www': 'website',
        'barrierefreiheit': 'accessibility', 'bemerkungen': 'notes',
        'lat': 'latitude', 'lng': 'longitude'
    }
    return df.rename(columns=mapping)

print("Helper functions defined ✓")

Helper functions defined ✓


## Step 3: Extract and Process Official Sources

In [51]:
print("\n🔧 Processing Official Sources")
print("-" * 70)

# Extract official sources
df_christmas = standardize_columns(extract_geojson_to_df(christmas_data, 'christmas_market'))
print(f"✓ Extracted {len(df_christmas)} current Christmas markets")

df_weekly_all = standardize_columns(extract_geojson_to_df(weekly_data, 'weekly_market'))

# Filter out flea markets
flea_keywords = ['floh', 'trödel', 'antik', 'sammler', 'kiefi', 'kinderfloh']
pattern = '|'.join(flea_keywords)
is_flea = df_weekly_all.get('market_name', pd.Series()).str.lower().str.contains(pattern, na=False)
df_weekly = df_weekly_all[~is_flea].copy()
print(f"✓ Filtered out {is_flea.sum()} flea markets")

# Combine official sources
df_official = pd.concat([df_christmas, df_weekly], ignore_index=True)
print(f"✓ Combined official sources: {len(df_official)} markets")


🔧 Processing Official Sources
----------------------------------------------------------------------
✓ Extracted 46 current Christmas markets
✓ Filtered out 12 flea markets
✓ Combined official sources: 75 markets


## Step 4: Process OSM Data

In [52]:
print("\n🌐 Processing OSM Data")
print("-" * 70)

# Extract OSM data
osm_records = []
for element in osm_data['elements']:
    record = {
        'market_name': element.get('tags', {}).get('name', ''),
        'latitude': element.get('lat', element.get('center', {}).get('lat', None)),
        'longitude': element.get('lon', element.get('center', {}).get('lon', None)),
        'opening_hours': element.get('tags', {}).get('opening_hours', ''),
        'website': element.get('tags', {}).get('website', ''),
        'data_source': 'OSM',
        'market_type': 'christmas_market' if 'xmas:feature' in element.get('tags', {})
                      else 'food_court' if element.get('tags', {}).get('amenity') == 'food_court'
                      else 'weekly_market'
    }
    osm_records.append(record)
df_osm = pd.DataFrame(osm_records)

print(f"✓ Extracted {len(df_osm)} markets from OSM")


🌐 Processing OSM Data
----------------------------------------------------------------------
✓ Extracted 147 markets from OSM


## Step 5: Deduplicate Across Sources

In [53]:
print("\n🔍 Deduplicating")
print("-" * 70)

def clean_name(name):
    if pd.isna(name): return ''
    name = str(name).lower()
    name = re.sub(r'\b(wochenmarkt|markt|market|weekly|ökomarkt)\b', '', name)
    return re.sub(r'\s+', ' ', name).strip()

df_official['name_clean'] = df_official['market_name'].apply(clean_name)
df_osm['name_clean'] = df_osm['market_name'].apply(clean_name)

# Keep only OSM markets not in official sources
osm_new = df_osm[~df_osm['name_clean'].isin(df_official['name_clean']) & (df_osm['name_clean'].str.len() > 0)]
print(f"✓ Found {len(osm_new)} additional markets from OSM")


🔍 Deduplicating
----------------------------------------------------------------------
✓ Found 110 additional markets from OSM


## Step 6: Combine All Sources

In [54]:
print("\n🔗 Combining All Sources")
print("-" * 70)

def ensure_columns(df, columns):
    for col in columns:
        if col not in df.columns:
            df[col] = ''
    return df[columns]

df_combined = pd.concat([
    ensure_columns(df_official, FINAL_COLUMNS),
    ensure_columns(osm_new, FINAL_COLUMNS)
], ignore_index=True)

print(f"✓ Total combined: {len(df_combined)} markets")


🔗 Combining All Sources
----------------------------------------------------------------------
✓ Total combined: 185 markets


## Step 7: Data Cleaning

In [55]:
print("\n🧹 Cleaning Data")
print("-" * 70)

# Keep only markets with coordinates
df_combined['latitude'] = pd.to_numeric(df_combined['latitude'], errors='coerce')
df_combined['longitude'] = pd.to_numeric(df_combined['longitude'], errors='coerce')
df_clean = df_combined.dropna(subset=['latitude', 'longitude']).copy()
print(f"✓ Markets with valid coordinates: {len(df_clean)}")

# Fill missing values
fill_values = {
    'market_name': 'Unknown Market', 'market_type': 'weekly_market',
    'address': 'Not specified', 'postal_code': 'Not available',
    'district': 'Unknown', 'opening_days': 'See website',
    'opening_hours': 'See website', 'operator': 'Not specified',
    'contact_email': 'Not available', 'website': 'Not available',
    'accessibility': 'Not specified', 'data_source': 'unknown',
    'notes': 'None'
}

for col, value in fill_values.items():
    if col in df_clean.columns:
        df_clean[col] = df_clean[col].fillna(value).replace('', value)

# Remove duplicates
df_clean['lat_rounded'] = df_clean['latitude'].round(3)
df_clean['lon_rounded'] = df_clean['longitude'].round(3)
before = len(df_clean)
df_clean = df_clean.drop_duplicates(subset=['market_name', 'lat_rounded', 'lon_rounded'], keep='first')
print(f"✓ Removed {before - len(df_clean)} duplicates")

df_clean = df_clean.drop(columns=['lat_rounded', 'lon_rounded', 'name_clean'], errors='ignore')


🧹 Cleaning Data
----------------------------------------------------------------------
✓ Markets with valid coordinates: 185
✓ Removed 0 duplicates


## Step 8: Spatial Join for Neighborhood and District IDs

This is the crucial step! We perform a spatial join to match market coordinates with Berlin administrative boundaries.

In [56]:
print("\n🗺️  Performing Spatial Join")
print("-" * 70)

# Create GeoDataFrame from market coordinates
df_clean['geometry'] = df_clean.apply(
    lambda row: Point(float(row['longitude']), float(row['latitude'])),
    axis=1
)
markets_gdf = gpd.GeoDataFrame(df_clean, geometry='geometry', crs='EPSG:4326')
print(f"✓ Created GeoDataFrame with {len(markets_gdf)} markets")

# Load Berlin neighborhoods (LOR Ortsteile)
lor_path = Path('mapping/lor_ortsteile.geojson')
if not lor_path.exists():
    lor_path = Path('food_markets/sources/lor_ortsteile.geojson')

berlin_gdf = gpd.read_file(lor_path)
print(f"✓ Loaded {len(berlin_gdf)} Berlin neighborhoods")

# Perform spatial join - use spatial_name as neighborhood_id!
markets_with_location = gpd.sjoin(
    markets_gdf,
    berlin_gdf[['BEZIRK', 'spatial_name', 'geometry']],
    how='left',
    predicate='within'
)

matched = markets_with_location['spatial_name'].notna().sum()
unmatched = markets_with_location['spatial_name'].isna().sum()
print(f"✓ Spatial join complete: {matched} matched, {unmatched} outside Berlin")

# Map district IDs using official 8-digit codes
markets_with_location['district_id'] = markets_with_location['BEZIRK'].map(DISTRICT_MAPPING)

# Fallback: try original district column
mask = markets_with_location['district_id'].isna()
markets_with_location.loc[mask, 'district_id'] = \
    markets_with_location.loc[mask, 'district'].map(DISTRICT_MAPPING)

# Update district name from BEZIRK (official spatial join result)
markets_with_location.loc[markets_with_location['BEZIRK'].notna(), 'district'] = \
    markets_with_location.loc[markets_with_location['BEZIRK'].notna(), 'BEZIRK']

# Set neighborhood_id from spatial_name (e.g., "0101", "0102")
markets_with_location['neighborhood_id'] = markets_with_location['spatial_name']

# IMPORTANT: Ensure neighborhood_id is always 4 characters with leading zeros
# This prevents pandas from treating it as an integer and dropping leading zeros
def format_neighborhood_id(val):
    if pd.isna(val):
        return val
    # Convert to string and ensure 4 characters with leading zeros
    return str(val).zfill(4)

markets_with_location['neighborhood_id'] = markets_with_location['neighborhood_id'].apply(format_neighborhood_id)

# Handle markets outside Berlin
markets_with_location['neighborhood_id'] = markets_with_location['neighborhood_id'].fillna('Outside Berlin')
markets_with_location['district_id'] = markets_with_location['district_id'].fillna('99999999')

berlin_markets = (markets_with_location['district_id'] != '99999999').sum()
print(f"✓ Final: {berlin_markets} markets in Berlin, {len(markets_with_location) - berlin_markets} outside")


🗺️  Performing Spatial Join
----------------------------------------------------------------------
✓ Created GeoDataFrame with 185 markets
✓ Loaded 96 Berlin neighborhoods
✓ Spatial join complete: 166 matched, 19 outside Berlin
✓ Final: 166 markets in Berlin, 19 outside


## Step 9: Fill Missing Addresses with Nominatim

Use reverse geocoding to complete missing addresses based on coordinates.

We enrich two types of addresses:
1. Completely missing addresses ('Not specified')
2. Addresses without house numbers (detected via regex `\d`)

Many official sources only provide street names like "Winterfeldtplatz" without house numbers. By checking for the absence of digits, we can identify and enrich these incomplete addresses.

**Important Note on Address Quality:**
Even after reverse geocoding, some addresses may still lack house numbers. This is **expected behavior** because:
- Markets are typically located in public spaces (plazas, squares, parks, or along streets)
- These locations don't have traditional building addresses with house numbers
- Examples: "Maybachufer" (embankment), "Helene-Weigel-Platz" (plaza), "Karl-Marx-Allee" (avenue)
- Nominatim returns the best available address for the coordinate - if no house number exists, only the street/plaza name is provided
- Approximately 50% of markets will have street names only, which is correct for their location type

In [57]:
print("\n🌐 Geocoding Missing Addresses")
print("-" * 70)

# Initialize Nominatim geocoder
geolocator = Nominatim(user_agent="berlin_food_markets_transformation")

def reverse_geocode_address(lat, lon):
    """Get address AND postal code from coordinates using Nominatim"""
    try:
        time.sleep(1)  # Respect Nominatim usage policy (max 1 request per second)
        location = geolocator.reverse(f"{lat}, {lon}", language='de')
        if location:
            # Extract street address and postal code from the result
            addr = location.raw.get('address', {})
            road = addr.get('road', '')
            house_number = addr.get('house_number', '')
            postcode = addr.get('postcode', '')
            
            address = f"{road} {house_number}".strip() if road else location.address.split(',')[0]
            return address, postcode
        return None, None
    except (GeocoderTimedOut, GeocoderServiceError):
        return None, None
    except Exception:
        return None, None

# Find addresses needing enrichment - BUT ONLY FOR BERLIN MARKETS
berlin_markets_mask = markets_with_location['district_id'] != '99999999'

# 1. Addresses that are 'Not specified' or NaN
# 2. Addresses that don't contain any numbers (e.g., just street name)
missing_address = (markets_with_location['address'] == 'Not specified') | (markets_with_location['address'].isna())
no_house_number = markets_with_location['address'].str.contains(r'\d', regex=True, na=False) == False

# Combine: addresses that need geocoding AND are in Berlin
needs_geocoding = (missing_address | no_house_number) & berlin_markets_mask
geocoding_count = needs_geocoding.sum()

print(f"Addresses completely missing: {missing_address.sum()}")
print(f"Addresses without house numbers: {no_house_number.sum()}")
print(f"Total addresses to geocode (Berlin only): {geocoding_count}")

if geocoding_count > 0:
    print(f"Geocoding addresses (this may take a while, 1 sec per request)...")
    filled = 0
    for idx in markets_with_location[needs_geocoding].index:
        lat = markets_with_location.loc[idx, 'latitude']
        lon = markets_with_location.loc[idx, 'longitude']
        
        if pd.notna(lat) and pd.notna(lon):
            address, postcode = reverse_geocode_address(lat, lon)
            if address:
                markets_with_location.loc[idx, 'address'] = address
                filled += 1
            if postcode:
                markets_with_location.loc[idx, 'postal_code'] = postcode
            if filled % 10 == 0:
                print(f"  Geocoded {filled}/{geocoding_count} addresses...")
    
    print(f"✓ Filled/enriched {filled} addresses using Nominatim reverse geocoding")
    
    # Check how many still lack house numbers after geocoding
    still_no_numbers = markets_with_location['address'].str.contains(r'\d', regex=True, na=False) == False
    print(f"ℹ️  Note: {still_no_numbers.sum()} addresses still lack house numbers (expected for public spaces)")
else:
    print(f"✓ All markets have complete addresses!")


🌐 Geocoding Missing Addresses
----------------------------------------------------------------------
Addresses completely missing: 110
Addresses without house numbers: 140
Total addresses to geocode (Berlin only): 135
Geocoding addresses (this may take a while, 1 sec per request)...
  Geocoded 10/135 addresses...
  Geocoded 20/135 addresses...
  Geocoded 30/135 addresses...
  Geocoded 40/135 addresses...


KeyboardInterrupt: 

## Step 10: Generate Primary Keys

Create unique market_id for each market (FM001, FM002, etc.).

In [ ]:
print("\n🆔 Generating Primary Keys")
print("-" * 70)

# Generate unique market_id for each market
markets_with_location['market_id'] = [f"FM{str(i+1).zfill(3)}" for i in range(len(markets_with_location))]
print(f"✓ Generated {len(markets_with_location)} unique market IDs (FM001 - FM{len(markets_with_location):03d})")


🆔 Generating Primary Keys
----------------------------------------------------------------------
✓ Generated 185 unique market IDs (FM001 - FM185)


## Step 11: Export to CSV

In [ ]:
print("\n💾 Exporting to CSV")
print("-" * 70)

# ONLY export Berlin markets (exclude outside Berlin)
berlin_markets_only = markets_with_location[markets_with_location['district_id'] != '99999999']
print(f"Filtering to Berlin markets only: {len(berlin_markets_only)} out of {len(markets_with_location)}")

# Select final columns
df_export = berlin_markets_only[FINAL_COLUMNS].copy()

# IMPORTANT: Validate that district_id is never empty (should be guaranteed by spatial join)
empty_district_ids = df_export['district_id'].isna().sum()
if empty_district_ids > 0:
    print(f"⚠️  WARNING: {empty_district_ids} markets have empty district_id!")
else:
    print(f"✓ All {len(df_export)} markets have district_id assigned")

# Fill missing values for all columns EXCEPT district_id and neighborhood_id
# (these should already be populated from spatial join)
for col in df_export.columns:
    if col not in ['district_id', 'neighborhood_id', 'market_id']:
        df_export[col] = df_export[col].fillna('Not available').replace('', 'Not available')

df_export.to_csv(OUTPUT_FILE, index=False, encoding='utf-8', na_rep='Not available')

print(f"✓ Exported to: {OUTPUT_FILE}")
print(f"✓ Total markets: {len(df_export)}")
print(f"✓ Columns: {len(df_export.columns)}")


💾 Exporting to CSV
----------------------------------------------------------------------
Filtering to Berlin markets only: 166 out of 185
✓ All 166 markets have district_id assigned
✓ Exported to: food_markets/berlin_food_markets_clean.csv
✓ Total markets: 166
✓ Columns: 18


## Step 12: Data Quality Validation

In [ ]:
print("\n✅ Validating Data Quality")
print("-" * 70)

# Verify no missing values
verify_df = pd.read_csv(OUTPUT_FILE, keep_default_na=False, na_values=[])
missing = verify_df.isnull().sum().sum()
print(f"✓ Missing values: {missing}")

# Verify required columns exist
required = ['market_id', 'neighborhood_id', 'district_id']
for col in required:
    if col in verify_df.columns:
        print(f"✓ {col} column present")
    else:
        print(f"✗ {col} column MISSING!")

# Verify district_id integrity
# IMPORTANT: Convert district_id to string for comparison (pandas reads it as int64)
verify_df['district_id'] = verify_df['district_id'].astype(str)
valid_district_ids = list(DISTRICT_MAPPING.values())  # Only Berlin districts (no 99999999)
invalid_districts = verify_df[~verify_df['district_id'].isin(valid_district_ids)]
if len(invalid_districts) > 0:
    print(f"⚠️  WARNING: {len(invalid_districts)} markets have invalid district_id!")
else:
    print(f"✓ All district_ids are valid (12 Berlin districts)")

# Verify all markets are in Berlin
berlin_count = verify_df['district_id'].isin(valid_district_ids).sum()
print(f"✓ All {berlin_count} markets are in Berlin (no outside markets exported)")

print("\n" + "=" * 70)
print("✨ TRANSFORMATION COMPLETE!")
print("=" * 70)


✅ Validating Data Quality
----------------------------------------------------------------------
✓ Missing values: 0
✓ market_id column present
✓ neighborhood_id column present
✓ district_id column present
✓ All district_ids are valid (12 Berlin districts)
✓ All 166 markets are in Berlin (no outside markets exported)

✨ TRANSFORMATION COMPLETE!


## Final Summary

In [ ]:
print(f"\n📊 FINAL SUMMARY:")
print(f"- Total markets: {len(verify_df)}")
print(f"- Market types: {verify_df['market_type'].value_counts().to_dict()}")
print(f"- All markets are in Berlin (outside markets excluded)")
print(f"- Data sources: {verify_df['data_source'].value_counts().to_dict()}")
print(f"- File size: {Path(OUTPUT_FILE).stat().st_size / 1024:.1f} KB")

# Show sample
print(f"\nSample data:")
verify_df.head(3)


📊 FINAL SUMMARY:
- Total markets: 166
- Market types: {'weekly_market': 113, 'christmas_market': 42, 'food_court': 11}
- All markets are in Berlin (outside markets excluded)
- Data sources: {'OSM': 110, 'berlin.de_official': 56}
- File size: 60.0 KB

Sample data:


,market_id,market_name,market_type,address,postal_code,district,neighborhood_id,district_id,opening_days,opening_hours,operator,contact_email,website,accessibility,latitude,longitude,data_source,notes
0,FM001,Winter am THF,christmas_market,Tempelhofer Damm 53,12101,Tempelhof-Schöneberg,703,11007007,See website,2. Adventswochenende: 05. bis 07.12.2025 \n3. ...,Flughafen Tempelhof (Tempelhof Projekt GmbH),kommunikation@thf-berlin.de,https://www.thf-berlin.de/winter-am-thf,ja,52.477,13.386,berlin.de_official,Die Alte Feuerwache ist barrierefrei von außen...
1,FM002,LGBT*Winterdays und Christmas Avenue Nollendor...,christmas_market,Kleiststraße,10787,Tempelhof-Schöneberg,701,11007007,See website,LGBTQIA* Winterdays:,Rutwiess Events Berlin & LGBT*Aktionsgemeinsch...,Info@Rutwiess-events.de,https://www.christmas-avenue.berlin/,nein,52.500,13.353,berlin.de_official,auf dem Markt sind Holzhecksel gestreut. Sollt...
2,FM003,30. Weihnachtsmarkt auf Lehmanns Bauernhof,christmas_market,Lehmanns Bauernhof \nAlt-Marienfelde 35-37,12277,Tempelhof-Schöneberg,705,11007007,See website,Freitag: 14 - 20 Uhr \nSamstag: 12 - 20 Uhr \n...,Lehmanns Bauernhof E&E Trade Event UG (haftung...,karstenemail@gmx.de,https://www.lehmannsbauernhof.de,Ja,52.413,13.366,berlin.de_official,"ebenerdiges Gelände, teilweise Kopfsteinpflast..."


## Next Steps

The dataset is now ready for **Step 3: Database Integration**

Required database schema:
- **Primary Key:** `market_id` (FM001, FM002, etc.)
- **Foreign Keys:** 
  - `district_id` → references `berlin_source_data.districts(district_id)`
  - `neighborhood_id` → spatial codes (0101, 0102, etc.)
- **Constraints:** NOT NULL on district_id and market_name

---

# Part 2: Database Integration

Now that we have a clean, validated dataset, we'll load it into the PostgreSQL database with proper schema and constraints.

## Database Configuration

**IMPORTANT:** Update the database credentials below before running this section.

In [ ]:
# Database configuration (SQLAlchemy)
from sqlalchemy import create_engine, text
import sqlalchemy

# Neon PostgreSQL connection string
# Format: postgresql://user:password@host:port/database?sslmode=require
DATABASE_URL = "postgresql://:@localhost:5433/?sslmode=require"
# UPDATE THIS with your actual Neon credentials

# Table configuration
SCHEMA = 'berlin_source_data'
TABLE = 'food_markets'

print(f"Target: {SCHEMA}.{TABLE}")

Target: berlin_source_data.food_markets


## Pre-Insert Validation

Quick validation checks before database insertion.

In [85]:
print("\n🔍 Pre-Insert Validation")
print("-" * 70)

# Reload CSV with proper dtypes
df = pd.read_csv(OUTPUT_FILE, dtype={'neighborhood_id': str, 'district_id': str})

# Check NOT NULL columns
required = ['market_id', 'market_name', 'district_id', 'latitude', 'longitude']
for col in required:
    null_count = df[col].isna().sum()
    status = "✓" if null_count == 0 else "❌"
    print(f"{status} {col}: {null_count} nulls")

# Check ID formats
print(f"\n✓ neighborhood_id: {(df['neighborhood_id'].str.len() == 4).all()} (all 4 chars)")
print(f"✓ district_id: {(df['district_id'].str.len() == 8).all()} (all 8 chars)")
print(f"✓ Unique district_ids: {df['district_id'].unique().tolist()}")

print(f"\n✓ Ready to insert {len(df)} markets")


🔍 Pre-Insert Validation
----------------------------------------------------------------------
✓ market_id: 0 nulls
✓ market_name: 0 nulls
✓ district_id: 0 nulls
✓ latitude: 0 nulls
✓ longitude: 0 nulls

✓ neighborhood_id: True (all 4 chars)
✓ district_id: True (all 8 chars)
✓ Unique district_ids: ['11007007', '11001001', '11010010', '11004004', '11011011', '11002002', '11006006', '11008008', '11009009', '11012012', '11003003', '11005005']

✓ Ready to insert 166 markets


## Database Connection & Schema Setup

In [86]:
print("\n🔌 Connecting to Database")
print("-" * 70)

# Create SQLAlchemy engine
engine = create_engine(DATABASE_URL)

# Test connection and create schema
with engine.connect() as conn:
    # Check PostgreSQL version
    result = conn.execute(text("SELECT version();"))
    version = result.scalar()
    print(f"✓ Connected: {version.split(',')[0]}")
    
    # # Create schema if not exists
    # conn.execute(text(f"CREATE SCHEMA IF NOT EXISTS {SCHEMA};"))
    # conn.commit()
    # print(f"✓ Schema '{SCHEMA}' ready")


🔌 Connecting to Database
----------------------------------------------------------------------
✓ Connected: PostgreSQL 16.9 on x86_64-pc-linux-gnu


## Create Table with Constraints

Schema includes PRIMARY KEY, NOT NULL, FOREIGN KEY, and CHECK constraints.

In [87]:
print("\n🏗️  Creating Table")
print("-" * 70)

create_table_sql = f"""
CREATE TABLE IF NOT EXISTS {SCHEMA}.{TABLE} (
    market_id VARCHAR(10) PRIMARY KEY,
    market_name VARCHAR(255) NOT NULL,
    market_type VARCHAR(50),
    address VARCHAR(255),
    postal_code VARCHAR(50),
    district VARCHAR(100),
    neighborhood_id VARCHAR(4) CHECK (LENGTH(neighborhood_id) = 4),
    district_id VARCHAR(8) NOT NULL,
    opening_days TEXT,
    opening_hours TEXT,
    operator VARCHAR(255),
    contact_email VARCHAR(255),
    website VARCHAR(500),
    accessibility TEXT,
    latitude DECIMAL(10, 7) NOT NULL,
    longitude DECIMAL(10, 7) NOT NULL,
    data_source VARCHAR(50),
    notes TEXT,
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
    
    CONSTRAINT fk_food_markets_district FOREIGN KEY (district_id)
        REFERENCES {SCHEMA}.districts(district_id)
        ON DELETE RESTRICT
        ON UPDATE CASCADE
);
"""

with engine.connect() as conn:
    # Drop existing table for clean re-runs
    conn.execute(text(f"DROP TABLE IF EXISTS {SCHEMA}.{TABLE} CASCADE;"))
    # Create table
    conn.execute(text(create_table_sql))
    conn.commit()

print(f"✓ Table '{SCHEMA}.{TABLE}' created")
print(f"  • PRIMARY KEY: market_id")
print(f"  • NOT NULL: market_name, district_id, latitude, longitude")
print(f"  • CHECK: neighborhood_id LENGTH = 4")
print(f"  • FOREIGN KEY: district_id → {SCHEMA}.districts")
print(f"    - ON DELETE RESTRICT (prevents orphan records)")
print(f"    - ON UPDATE CASCADE (maintains referential integrity)")


🏗️  Creating Table
----------------------------------------------------------------------
✓ Table 'berlin_source_data.food_markets' created
  • PRIMARY KEY: market_id
  • NOT NULL: market_name, district_id, latitude, longitude
  • CHECK: neighborhood_id LENGTH = 4
  • FOREIGN KEY: district_id → berlin_source_data.districts
    - ON DELETE RESTRICT (prevents orphan records)
    - ON UPDATE CASCADE (maintains referential integrity)


## Insert Data

In [88]:
print("\n📤 Inserting Data")
print("-" * 70)

# Select columns to insert (exclude created_at, it has DEFAULT)
columns = [
    'market_id', 'market_name', 'market_type', 'address', 'postal_code',
    'district', 'neighborhood_id', 'district_id', 'opening_days', 'opening_hours',
    'operator', 'contact_email', 'website', 'accessibility',
    'latitude', 'longitude', 'data_source', 'notes'
]

# Insert using pandas to_sql
df[columns].to_sql(
    name=TABLE,
    con=engine,
    schema=SCHEMA,
    if_exists='append',
    index=False,
    method='multi'
)

print(f"✓ Inserted {len(df)} markets")

# Verify
with engine.connect() as conn:
    result = conn.execute(text(f"SELECT COUNT(*) FROM {SCHEMA}.{TABLE};"))
    count = result.scalar()
    print(f"✓ Verified: {count} rows in table")


📤 Inserting Data
----------------------------------------------------------------------
✓ Inserted 166 markets
✓ Verified: 166 rows in table


## Validation & Summary

In [89]:
print("\n✅ Validation")
print("-" * 70)

with engine.connect() as conn:
    # Count by type
    result = conn.execute(text(f"""
        SELECT market_type, COUNT(*) 
        FROM {SCHEMA}.{TABLE} 
        GROUP BY market_type 
        ORDER BY COUNT(*) DESC;
    """))
    print("Markets by type:")
    for row in result:
        print(f"  • {row[0]}: {row[1]}")
    
    # Check foreign key integrity
    result = conn.execute(text(f"""
        SELECT COUNT(*) 
        FROM {SCHEMA}.{TABLE} fm
        JOIN {SCHEMA}.districts d ON fm.district_id = d.district_id;
    """))
    fk_count = result.scalar()
    print(f"\n✓ All {fk_count} markets have valid district references")

print(f"\n{'='*70}")
print("✅ DATABASE INTEGRATION COMPLETE!")
print(f"{'='*70}")


✅ Validation
----------------------------------------------------------------------
Markets by type:
  • weekly_market: 113
  • christmas_market: 42
  • food_court: 11

✓ All 166 markets have valid district references

✅ DATABASE INTEGRATION COMPLETE!


---

# Documentation

## Pipeline Overview

This notebook implements an end-to-end pipeline for Berlin food & weekly markets data:
1. **Fetch** official data from Berlin.de APIs
2. **Transform** and clean data with spatial joins
3. **Validate** data quality and referential integrity
4. **Load** into PostgreSQL with proper constraints

## Data Sources

**Primary:** Berlin.de Official GeoJSON
- Christmas Markets: https://www.berlin.de/sen/web/service/maerkte-feste/weihnachtsmaerkte/index.php/index/all.geojson?q=
- Weekly/Flea Markets: https://www.berlin.de/sen/web/service/maerkte-feste/wochen-troedelmaerkte/index.php/index/all.geojson?q=

**Supplemental:**
- OpenStreetMap (gap filling for markets not in official data)
- LOR Ortsteile GeoJSON (Berlin neighborhood boundaries for spatial join)

## Key Transformation Steps

1. **Extract & Standardize**: Load GeoJSON, standardize column names, filter past Christmas markets
2. **Deduplicate**: Cross-reference OSM against official sources, keep unique markets only
3. **Spatial Join**: Match coordinates to neighborhoods, extract 4-char `neighborhood_id` and 8-char `district_id`
4. **Address Enrichment**: Reverse geocode missing addresses using Nominatim
5. **Export**: Generate unique `market_id` (FM001, FM002...), export to CSV

## Table: food_markets

### Column Specifications

| Column | Type | Constraints | Description |
|--------|------|-------------|-------------|
| `market_id` | VARCHAR(10) | PRIMARY KEY | Unique market identifier (FM001-FM999) |
| `market_name` | VARCHAR(255) | NOT NULL | Official market name |
| `market_type` | VARCHAR(50) | | Type: 'weekly_market', 'christmas_market', 'food_court' |
| `address` | VARCHAR(255) | | Street address (may lack house numbers for public spaces) |
| `postal_code` | VARCHAR(50) | | Berlin postal code (5 digits) |
| `district` | VARCHAR(100) | | District name (Berlin administrative district) |
| `neighborhood_id` | VARCHAR(4) | CHECK LENGTH=4 | 4-char neighborhood code with leading zeros (0101-1211) |
| `district_id` | VARCHAR(8) | NOT NULL, FK | 8-digit district code (11001001-11012012) |
| `opening_days` | TEXT | | Days of operation |
| `opening_hours` | TEXT | | Operating hours |
| `operator` | VARCHAR(255) | | Market operator/organizer |
| `contact_email` | VARCHAR(255) | | Contact email address |
| `website` | VARCHAR(500) | | Market website URL |
| `accessibility` | TEXT | | Accessibility information |
| `latitude` | DECIMAL(10,7) | NOT NULL | GPS latitude (52.xxx) |
| `longitude` | DECIMAL(10,7) | NOT NULL | GPS longitude (13.xxx) |
| `data_source` | VARCHAR(50) | | Source: 'berlin.de_official', 'OSM' |
| `notes` | TEXT | | Additional notes and comments |
| `created_at` | TIMESTAMP | DEFAULT NOW() | Record creation timestamp |

### Constraints

#### Primary Key
- **market_id**: Ensures each market has a unique identifier
- Format: FM001, FM002, ..., FM999
- Sequential assignment during transformation

#### Foreign Key
```sql
CONSTRAINT fk_food_markets_district 
    FOREIGN KEY (district_id)
    REFERENCES berlin_source_data.districts(district_id)
    ON DELETE RESTRICT
    ON UPDATE CASCADE
```

**Rationale:**
- **ON DELETE RESTRICT**: Prevents deletion of districts that have associated markets, protecting data integrity
- **ON UPDATE CASCADE**: Automatically updates food_markets.district_id if districts.district_id changes, maintaining referential integrity

#### Check Constraint
```sql
CHECK (LENGTH(neighborhood_id) = 4)
```

**Rationale:**
- Enforces 4-character format with leading zeros (e.g., "0703", not "703")
- Prevents data quality issues from pandas/CSV treating IDs as integers
- Aligns with Berlin's official LOR (Lebensweltlich Orientierte Räume) neighborhood coding system

#### NOT NULL Constraints
- **market_name**: Essential for identification
- **district_id**: Required for foreign key relationship and geographic analysis
- **latitude, longitude**: Core geographic data for spatial queries and mapping

### Neighborhood ID Format
The 4-character neighborhood_id with leading zeros is **critical**:
- Correct: "0703", "0101", "1211"
- Incorrect: "703", "101", "1211" (missing leading zeros)

The CHECK constraint enforces this at the database level, and the transformation pipeline uses `str.zfill(4)` to ensure proper formatting.

### Foreign Key Integrity
All 166 markets have valid `district_id` references to the `districts` table, verified during insertion. The ON DELETE RESTRICT constraint prevents accidental deletion of districts that have associated markets.

## Sample Queries

### Count markets by district
```sql
SELECT 
    d.district_name,
    COUNT(fm.market_id) as market_count
FROM berlin_source_data.food_markets fm
JOIN berlin_source_data.districts d ON fm.district_id = d.district_id
GROUP BY d.district_name
ORDER BY market_count DESC;
```

### Find markets by type in a specific district
```sql
SELECT 
    market_name,
    market_type,
    address,
    opening_days
FROM berlin_source_data.food_markets
WHERE district_id = '11007007'  -- Tempelhof-Schöneberg
  AND market_type = 'weekly_market'
ORDER BY market_name;
```

### Markets within a neighborhood
```sql
SELECT 
    market_name,
    address,
    latitude,
    longitude
FROM berlin_source_data.food_markets
WHERE neighborhood_id = '0703'  -- Tempelhof neighborhood
ORDER BY market_name;
```

### Markets with accessibility information
```sql
SELECT 
    market_name,
    district,
    accessibility,
    website
FROM berlin_source_data.food_markets
WHERE accessibility != 'Not specified'
  AND accessibility IS NOT NULL
ORDER BY district, market_name;
```

## Key Assumptions

1. **Coordinate Accuracy**: Trust official source coordinates; markets outside Berlin excluded
2. **Temporal Validity**: Christmas markets filtered by end date; no end date = current/recurring
3. **Address Completeness**: ~50% lack house numbers (markets in public spaces - plazas, parks)
4. **neighborhood_id Format**: CRITICAL - Must be 4 chars with leading zeros (enforced via `str.zfill(4)`)
5. **OSM Data**: Used only for gap-filling, not as primary source
```